shard.go

```diff
@@ -25,34 +25,48 @@ type cacheShard struct {
 	stats Stats
 }
 
-func (s *cacheShard) get(key string, hashedKey uint64) ([]byte, error) {
+func (s *cacheShard) getHelper(key string, hashedKey uint64) ([]byte, uint64, error) {
 	s.lock.RLock()
 	itemIndex := s.hashmap[hashedKey]
 
 	if itemIndex == 0 {
 		s.lock.RUnlock()
 		s.miss()
-		return nil, ErrEntryNotFound
+		return nil, 0, ErrEntryNotFound
 	}
 
 	wrappedEntry, err := s.entries.Get(int(itemIndex))
 	if err != nil {
 		s.lock.RUnlock()
 		s.miss()
-		return nil, err
+		return nil, 0, err
 	}
 	if entryKey := readKeyFromEntry(wrappedEntry); key != entryKey {
 		if s.isVerbose {
 			s.logger.Printf("Collision detected. Both %q and %q have the same hash %x", key, entryKey, hashedKey)
 		}
 		s.lock.RUnlock()
 		s.collision()
-		return nil, ErrEntryNotFound
+		return nil, 0, ErrEntryNotFound
 	}
 	entry := readEntry(wrappedEntry)
 	s.lock.RUnlock()
 	s.hit()
-	return entry, nil
+	return entry, readTimestampFromEntry(wrappedEntry), nil
```

Comment:
This could potentially decrease bigcache performance. We need to benchmark it. What I'd prefer is to create new function that will return timestamp and do not use it for regular `Get` just for `getWithInfo`

Reply:
How about this: `getHelper` returns wrapped entry and then `get` returns only datat or error and `getWithInfo` has all logic?

Reply:
yes i will generate benchmarking numbers for `get` & `getWithInfo` and returning wrapped entry makes sense.. Let me resend PR

Reply:
Number on Mac are as follows:

```
BenchmarkWriteToCacheWith1Shard-12                             	 2000000	       694 ns/op	     579 B/op	       3 allocs/op
BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard-12      	 3000000	       449 ns/op	      41 B/op	       3 allocs/op
BenchmarkWriteToUnlimitedCacheWithSmallInitSizeAnd1Shard-12    	 1000000	      1670 ns/op	    3448 B/op	       2 allocs/op
BenchmarkWriteToCache/1-shards-12                              	 2000000	       574 ns/op	     579 B/op	       3 allocs/op
BenchmarkWriteToCache/512-shards-12                            	10000000	       164 ns/op	     573 B/op	       3 allocs/op
BenchmarkWriteToCache/1024-shards-12                           	10000000	       151 ns/op	     574 B/op	       3 allocs/op
BenchmarkWriteToCache/8192-shards-12                           	10000000	       186 ns/op	     579 B/op	       3 allocs/op
BenchmarkReadFromCache/1-shards-12                             	 5000000	       324 ns/op	     271 B/op	       2 allocs/op
BenchmarkReadFromCache/512-shards-12                           	 5000000	       330 ns/op	     271 B/op	       2 allocs/op
BenchmarkReadFromCache/1024-shards-12                          	 5000000	       329 ns/op	     271 B/op	       2 allocs/op
BenchmarkReadFromCache/8192-shards-12                          	 5000000	       323 ns/op	     271 B/op	       2 allocs/op
BenchmarkReadFromCacheWithInfo/1-shards-12                     	 5000000	       326 ns/op	     271 B/op	       2 allocs/op
BenchmarkReadFromCacheWithInfo/512-shards-12                   	 5000000	       330 ns/op	     271 B/op	       2 allocs/op
BenchmarkReadFromCacheWithInfo/1024-shards-12                  	 5000000	       327 ns/op	     271 B/op	       2 allocs/op
BenchmarkReadFromCacheWithInfo/8192-shards-12                  	 5000000	       322 ns/op	     271 B/op	       2 allocs/op
BenchmarkIterateOverCache/512-shards-12                        	10000000	       244 ns/op	      20 B/op	       2 allocs/op
BenchmarkIterateOverCache/1024-shards-12                       	10000000	       237 ns/op	      20 B/op	       2 allocs/op
BenchmarkIterateOverCache/8192-shards-12                       	10000000	       220 ns/op	      20 B/op	       2 allocs/op
BenchmarkWriteToCacheWith1024ShardsAndSmallShardInitSize-12    	10000000	       355 ns/op	    1154 B/op	       3 allocs/op
BenchmarkReadFromCacheNonExistentKeys/1-shards-12              	10000000	       185 ns/op	       8 B/op	       0 allocs/op
BenchmarkReadFromCacheNonExistentKeys/512-shards-12            	10000000	       191 ns/op	       7 B/op	       0 allocs/op
BenchmarkReadFromCacheNonExistentKeys/1024-shards-12           	10000000	       189 ns/op	       7 B/op	       0 allocs/op
BenchmarkReadFromCacheNonExistentKeys/8192-shards-12           	10000000	       190 ns/op	       7 B/op	       0 allocs/op
BenchmarkFnvHashSum64-12                                       	200000000	         8.66 ns/op
BenchmarkFnvHashStdLibSum64-12                                 	50000000	        38.3 ns/op
PASS

```

Reply:
Can you use https://godoc.org/golang.org/x/tools/cmd/benchcmp

Reply:
Numbers for bench compare are as follows:
```

benchmark                                                       old allocs     new allocs     delta
BenchmarkWriteToCacheWith1Shard-12                              3              3              +0.00%
BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard-12       3              3              +0.00%
BenchmarkWriteToUnlimitedCacheWithSmallInitSizeAnd1Shard-12     2              2              +0.00%
BenchmarkWriteToCache/1-shards-12                               3              3              +0.00%
BenchmarkWriteToCache/512-shards-12                             3              3              +0.00%
BenchmarkWriteToCache/1024-shards-12                            3              3              +0.00%
BenchmarkWriteToCache/8192-shards-12                            3              3              +0.00%
BenchmarkReadFromCache/1-shards-12                              2              2              +0.00%
BenchmarkReadFromCache/512-shards-12                            2              2              +0.00%
BenchmarkReadFromCache/1024-shards-12                           2              2              +0.00%
BenchmarkReadFromCache/8192-shards-12                           2              2              +0.00%
BenchmarkIterateOverCache/512-shards-12                         2              2              +0.00%
BenchmarkIterateOverCache/1024-shards-12                        2              2              +0.00%
BenchmarkIterateOverCache/8192-shards-12                        2              2              +0.00%
BenchmarkWriteToCacheWith1024ShardsAndSmallShardInitSize-12     3              3              +0.00%
BenchmarkReadFromCacheNonExistentKeys/1-shards-12               0              0              +0.00%
BenchmarkReadFromCacheNonExistentKeys/512-shards-12             0              0              +0.00%
BenchmarkReadFromCacheNonExistentKeys/1024-shards-12            0              0              +0.00%
BenchmarkReadFromCacheNonExistentKeys/8192-shards-12            0              0              +0.00%

benchmark                                                       old bytes     new bytes     delta
BenchmarkWriteToCacheWith1Shard-12                              579           579           +0.00%
BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard-12       41            41            +0.00%
BenchmarkWriteToUnlimitedCacheWithSmallInitSizeAnd1Shard-12     3448          3448          +0.00%
BenchmarkWriteToCache/1-shards-12                               579           579           +0.00%
BenchmarkWriteToCache/512-shards-12                             574           573           -0.17%
BenchmarkWriteToCache/1024-shards-12                            574           574           +0.00%
BenchmarkWriteToCache/8192-shards-12                            579           579           +0.00%
BenchmarkReadFromCache/1-shards-12                              271           271           +0.00%
BenchmarkReadFromCache/512-shards-12                            271           271           +0.00%
BenchmarkReadFromCache/1024-shards-12                           271           271           +0.00%
BenchmarkReadFromCache/8192-shards-12                           271           271           +0.00%
BenchmarkIterateOverCache/512-shards-12                         20            20            +0.00%
BenchmarkIterateOverCache/1024-shards-12                        20            20            +0.00%
BenchmarkIterateOverCache/8192-shards-12                        20            20            +0.00%
BenchmarkWriteToCacheWith1024ShardsAndSmallShardInitSize-12     1154          1154          +0.00%
BenchmarkReadFromCacheNonExistentKeys/1-shards-12               7             7             +0.00%
BenchmarkReadFromCacheNonExistentKeys/512-shards-12             7             7             +0.00%
BenchmarkReadFromCacheNonExistentKeys/1024-shards-12            7             7             +0.00%
BenchmarkReadFromCacheNonExistentKeys/8192-shards-12            7             7             +0.00%

```